scipy/optimize/_quasi_newton_approx.py

```diff
@@ -0,0 +1,261 @@
+"""quasi-Newton Hessian approximations"""
+
+from __future__ import division, print_function, absolute_import
+import numpy as np
+from numpy.linalg import norm
+from scipy.linalg import get_blas_funcs
+
+
+class QuasiNewtonApprox(object):
+    """Quasi-Newton approximation."""
+
+    _syr = get_blas_funcs('syr', dtype='d')  # Symetric rank 1 update
+    _syr2 = get_blas_funcs('syr2', dtype='d')  # Symetric rank 2 update
+    _symv = get_blas_funcs('symv', dtype='d')  # Symetric matrix-vector product
+
+    def instanciate_matrix(self, delta_x, delta_grad, approx_type='hess'):
```

Comment:
About your second comment: You are correct, I choose to implement both, but I only need the Hessian. I choose to implement both because I thought it would be easier to test (since the Hessian can be compared with the inverse) and it was not much extra work. Furthermore, I believe it will be useful to have a common implementation of quasi-Newton updates to be shared among optimization methods.

For the `trust_constr` method, the Hessian is required by the QP problem, and hence I only use ``hess`` mode.

Reply:
> I think this method logically should be __init__ method. It should accept the approximation kind (hess or inv_hess) and the number of variables. Yes, you won't be able to pass BFGS object to the function, but you can pass class instead. I think it will be a better approach, think about it.

I will suggest an alternative approach and try to explain why I think it is better.

The problem here is  that there are some necessary arguments that we do not want to require the user to specify at the time he initializes an object instance. More specifically, the number of variables and which ``approx_type`` is required by the method being used.

I think the best solution in this case is to define two methods: ``__init__`` and ``initialize``. In this case all user options will be specified at ``__init__`` and the remaining options that are dependent on the method should be passed on to ``initialize`` internally.

For instance, for the BFGS method we would have:
```
def __init__(self, exception_strategy='skip_update', min_curvature=None, [other user options...]):
    pass

def initialize(self, n, approx_type):
    pass
```

This way the use could initialize an BFGS object, specifying all options. For instance:
```
BFGS(exception_strategy='damped_bfgs', min_curvature=1e-10)
```
and the remaining required information would be passed to the object internally.

I think this is better than to ask the user to pass a **class** to the function, since we would have to request an dictionary containing the options as well. Which I think would be less user friendly.

Reply:
> I think this is better than to ask the user to pass a class to the function, since we would have to request an dictionary containing the options as well. Which I think would be less user friendly.

That's for sure. But I thought that you don't need any user-defined options. Now I see that this is a flawed approach if we want to make this part open for modification and introduction of new strategies.

I think `initialize` is not really ideal name, but I don't know if there are better names.

Reply:
Ok, I will keep ``initialize`` for now. If we come up with something better later I will change it.